src/gwc/src/main/java/org/geoserver/gwc/config/GeoserverXMLResourceProvider.java

```diff
@@ -0,0 +1,178 @@
+        
+        this.configFileName = configFileName;
+        
+        if(configFileDirectory!=null) {
+            // Use the given path
+            if (configFileDirectory.startsWith("/") || configFileDirectory.contains(":\\")
+                    || configFileDirectory.startsWith("\\\\")) {
+                
+                log.info("Provided configuration directory as absolute path '" + configFileDirectory + "'");
+                this.configDirectory = Files.asResource(new File(configFileDirectory));
+            } else {
+                log.info("Provided configuration directory in the resource store. ");
+                this.configDirectory = resourceStore.get(configFileDirectory);
+            }
+        } else {
+            this.configDirectory = resourceStore.get(DEFAULT_CONFIGURATION_DIR_NAME);
+        }
+        log.info("Will look for geowebcache.xml in '" + configDirectory + "'");
+    }
+    
+    public GeoserverXMLResourceProvider(final String configFileName,
+            final ResourceStore resourceStore) throws ConfigurationException {
+        this(null, configFileName, resourceStore);        
+    }
+   
+
+    @Override
+    public InputStream in() throws IOException {
+        return findOrCreateConfFile().in();
+    }
+
+    @Override
+    public OutputStream out() throws IOException {
+        return findOrCreateConfFile().out();
+    }
+
+    @Override
+    public void backup() throws IOException {
+        backUpConfig(findOrCreateConfFile());
+    }
+
+    @Override
+    public String getId() {
+        return configDirectory.path();
+    }
+
+    @Override
+    public void setTemplate(final String templateLocation) {
+        this.templateLocation = templateLocation;
+    }
+
+    private Resource findConfigFile() throws IOException {
+        return configDirectory.get(configFileName);
+    }
+    
+    public String getLocation() throws IOException {
+        return findConfigFile().path();
+    }
+
+    private Resource findOrCreateConfFile() throws IOException {
+        Resource xmlFile = findConfigFile();
+
+        if (Resources.exists(xmlFile)) {
+            log.info("Found configuration file in " + configDirectory.path());
+        } else if (templateLocation != null) {
+            log.warn("Found no configuration file in config directory, will create one at '"
+                    + xmlFile.path() + "' from template "
+                    + getClass().getResource(templateLocation).toExternalForm());
+            // grab template from classpath
+            try {
+                 IOUtils.copy(getClass().getResourceAsStream(templateLocation), 
+                        xmlFile.out());
+            } catch (IOException e) {
+                throw new IOException("Error copying template config to "
+                        + xmlFile.path(), e);
+            }
+        } 
+
+        return xmlFile;
+    }
+    
+
+    private void backUpConfig(final Resource xmlFile) throws IOException {
```

Comment:
If you look at the LockProvider code you can see how we handled moving code from GWC to GeoServer.

GWC is "complicated" in that credit was often not recorded, so we had to check github permission for LockProvider and ask Andrea specifically for permission.